src/particles.h

```diff
@@ -20,19 +20,269 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+
+		T min, max;
+		f32 bias = 0;
+
+		RangedParameter() = default;
+		RangedParameter(const This& a)             : min(a.min), max(a.max) {};
+		RangedParameter(T _min, T _max)            : min(_min),  max(_max)  {};
+		template <typename M> RangedParameter(M b) : min(b),     max(b)     {};
+
+		// these functions handle the old range serialization "format"; bias must
+		// be manually encoded in a separate part of the stream. NEVER ADD FIELDS
+		// TO THESE FUNCTIONS
+		void legacySerialize(std::ostream& os) const {
+			min.serialize(os);
+			max.serialize(os);
+		}
+		void legacyDeSerialize(std::istream& is) {
+			min.deSerialize(is);
+			max.deSerialize(is);
+		}
+
+		// these functions handle the format used by new fields. new fields go here
+		void serialize(std::ostream &os) const {
+			legacySerialize(os);
+			writeF32(os, bias);
+		};
+		void deSerialize(std::istream &is) {
+			legacyDeSerialize(is);
+			bias = readF32(is);
+		};
+
+		This interpolate(float fac, const This against) const {
+			This r;
+			r.min = min.interpolate(fac, against.min);
+			r.max = max.interpolate(fac, against.max);
+			return r;
+		}
+
+		T pickWithin() {
+			typename T::pickFactors values;
+			auto p = numericAbsolute(bias) + 1;
+			for (size_t i = 0; i < sizeof(values) / sizeof(values[0]); ++i) {
+				if (bias < 0)
+					values[i] = 1.0f - pow(myrand_float(), p);
+				else
+					values[i] = pow(myrand_float(), p);
+			}
+			return T::pick(values, min, max);
+		}
+	};
+
+	enum class TweenStyle { fwd, rev, pulse, flicker };
+
+	template <typename T>
+	struct TweenedParameter : public Parameter {
+		using ValType = T;
+		using This = TweenedParameter<T>;
+
+		TweenStyle style = TweenStyle::fwd;
+		u16 reps = 1;
+		f32 beginning = 0.0f;
+
+		T start, end;
+
+		TweenedParameter() = default;
+		TweenedParameter(const This& a) = default;
+		TweenedParameter(T _start, T _end)          : start(_start),  end(_end) {};
+		template <typename M> TweenedParameter(M b) : start(b),       end(b) {};
+
+		T blend(float fac) const {
+			// warp time coordinates in accordance w/ settings
+			if (fac > beginning) {
+				// remap for beginning offset
+				auto len = 1 - beginning;
+				fac -= beginning;
+				fac /= len;
+
+				// remap for repetitions
+				fac *= reps;
+				if (fac > 1) // poor man's modulo
+					fac -= (decltype(reps))fac;
+
+				// remap for style
+				switch (style) {
+					case TweenStyle::fwd: /* do nothing */  break;
+					case TweenStyle::rev: fac = 1.0f - fac; break;
+					case TweenStyle::pulse:
+					case TweenStyle::flicker: {
+						if (fac > 0.5f) {
+							fac = 1.f - (fac*2.f - 1.f);
+						} else {
+							fac = fac * 2;
+						}
+						if (style == TweenStyle::flicker) {
+							fac *= myrand_range(0.7f, 1.0f);
+						}
+					}
+				}
+				if (fac>1.f)
+					fac = 1.f;
+				else if (fac<0.f)
+					fac = 0.f;
+			} else {
+				fac = (style == TweenStyle::rev) ? 1.f : 0.f;
+			}
+
+			return start.interpolate(fac, end);
+		}
```

Comment:
...this however should reside in a `.cpp` (rule of thumb is more than three lines)

Reply:
this is a template. it cannot go in a cpp file unless it is only used in that file. since this is a type that's used in the network protocol that's a nonstarter

Reply:
Oops missed that.